common/model/csvExport/CsvDataExportModel.js

```diff
@@ -36,53 +36,74 @@ const DEFAULT_OPTIONS = { includeCategoryItemsLabels: true, includeTaxonScientif
     this._initColumns()
   }
 
-  _initAttributeDefs() {
-    const { includeFiles } = this.options
-    let descendantDefs = NodeDef.isEntity(this.nodeDefContext)
-      ? Survey.getNodeDefDescendantAttributesInSingleEntities(this.nodeDefContext)(this.survey)
-      : [this.nodeDefContext] // Multiple attribute
-    if (!includeFiles) {
-      descendantDefs = descendantDefs.filter((nodeDef) => !NodeDef.isFile(nodeDef))
-    }
+  _initColumns() {
+    const { addCycle } = this.options
 
-    let parentKeys = []
-    Survey.visitAncestors(this.nodeDefContext, (n) => {
-      const keys = Survey.getNodeDefKeys(n)(this.survey)
-      parentKeys = parentKeys.concat(keys)
-    })(this.survey)
+const descendantAttributeColumns = this._extractAttributeDefsColumns()
+
+    const ancestorsKeyColumns = this._extractAncestorsKeysColumns()
 
-    this.attributeDefs = parentKeys.reverse().concat(descendantDefs)
+    this.columns = [
+      ...(addCycle ? [{ header: RECORD_CYCLE_HEADER }] : []),
+      ...ancestorsKeyColumns,
+      ...descendantAttributeColumns,
+    ]
   }
 
-  _initColumns() {
-    const { includeCategoryItemsLabels, includeTaxonScientificName, addCycle } = this.options
+  _createColumnsFromAttributeDefs({ attributeDefs }) {
+    const { includeCategoryItemsLabels, includeTaxonScientificName } = this.options
 
-    const columns = []
-    this.attributeDefs.forEach((nodeDef) => {
+    return attributeDefs.reduce((acc, nodeDef) => {
       const columnsGetter = columnsByNodeDefType[NodeDef.getType(nodeDef)]
       const columnsPerAttribute = columnsGetter
         ? columnsGetter({ nodeDef, includeCategoryItemsLabels, includeTaxonScientificName })
         : [getMainColumn({ nodeDef })]
-      columns.push(...columnsPerAttribute)
-    })
-    // Cycle is 0-based
-    this.columns = [...(addCycle ? [{ header: RECORD_CYCLE_HEADER }] : []), ...columns]
+      return [...acc, ...columnsPerAttribute]
+    }, [])
+  }
+
+  _extractAncestorsKeysColumns() {
+    const ancestorsKeyColumns = []
+
+    Survey.visitAncestors(this.nodeDefContext, (nodeDefAncestor) => {
+      const ancestorKeyDefs = Survey.getNodeDefKeys(nodeDefAncestor)(this.survey)
+      const ancestorKeyColumns = this._createColumnsFromAttributeDefs({
+        attributeDefs: ancestorKeyDefs,
+      })
+      ancestorsKeyColumns.unshift(...ancestorKeyColumns)
+    })(this.survey)
+
+    return ancestorsKeyColumns
+  }
+
+  _extractAttributeDefsColumns() {
+    const { includeFiles, includeAnalysis } = this.options
+
+    let descendantDefs = NodeDef.isEntity(this.nodeDefContext)
+      ? Survey.getNodeDefDescendantAttributesInSingleEntities(this.nodeDefContext)(this.survey)
+      : [this.nodeDefContext] // Multiple attribute
+
+    if (!includeFiles) {
+      descendantDefs = descendantDefs.filter((nodeDef) => !NodeDef.isFile(nodeDef))
+    }
+    if (!includeAnalysis) {
+      descendantDefs = descendantDefs.filter((nodeDef) => !NodeDef.isAnalysis(nodeDef))
+    }
```

Comment:
maybe you can put the filters inside only 1 filter to decrease the complexity

Reply:
right